source/ImdsClient.cpp

```diff
@@ -0,0 +1,401 @@
+            }
+
+            void ImdsClient::s_onVectorResourceAcquired(const aws_array_list *array, int errorCode, void *userData)
+            {
+                WrappedCallbackArgs<OnVectorResourceAcquired> *callbackArgs =
+                    static_cast<WrappedCallbackArgs<OnVectorResourceAcquired> *>(userData);
+                callbackArgs->callback(
+                    ArrayListToVector<ByteCursor, StringView>(array, ByteCursorToStringView),
+                    errorCode,
+                    callbackArgs->userData);
+                Aws::Crt::Delete(callbackArgs, callbackArgs->allocator);
+            }
+
+            void ImdsClient::s_onCredentialsAcquired(const aws_credentials *credentials, int errorCode, void *userData)
+            {
+                WrappedCallbackArgs<OnCredentialsAcquired> *callbackArgs =
+                    static_cast<WrappedCallbackArgs<OnCredentialsAcquired> *>(userData);
+                auto credentialsPtr = Aws::Crt::MakeShared<Auth::Credentials>(callbackArgs->allocator, credentials);
+                callbackArgs->callback(credentials, errorCode, callbackArgs->userData);
+                Aws::Crt::Delete(callbackArgs, callbackArgs->allocator);
+            }
+
+            void ImdsClient::s_onIamProfileAcquired(
+                const aws_imds_iam_profile *iamProfileInfo,
+                int errorCode,
+                void *userData)
+            {
+                WrappedCallbackArgs<OnIamProfileAcquired> *callbackArgs =
+                    static_cast<WrappedCallbackArgs<OnIamProfileAcquired> *>(userData);
+                IamProfile iamProfile;
+                iamProfile.lastUpdated = aws_date_time_as_epoch_secs(&(iamProfileInfo->last_updated));
+                iamProfile.instanceProfileArn = StringView(
+                    reinterpret_cast<char *>(iamProfileInfo->instance_profile_arn.ptr),
+                    iamProfileInfo->instance_profile_arn.len);
+                iamProfile.instanceProfileId = StringView(
+                    reinterpret_cast<char *>(iamProfileInfo->instance_profile_id.ptr),
+                    iamProfileInfo->instance_profile_id.len);
+                callbackArgs->callback(iamProfile, errorCode, callbackArgs->userData);
+                Aws::Crt::Delete(callbackArgs, callbackArgs->allocator);
+            }
+
+            void ImdsClient::s_onInstanceInfoAcquired(
+                const aws_imds_instance_info *instanceInfo,
+                int errorCode,
+                void *userData)
+            {
+                WrappedCallbackArgs<OnInstanceInfoAcquired> *callbackArgs =
+                    static_cast<WrappedCallbackArgs<OnInstanceInfoAcquired> *>(userData);
+                InstanceInfo info;
+                info.marketplaceProductCodes = ArrayListToVector<ByteCursor, StringView>(
+                    &(instanceInfo->marketplace_product_codes), ByteCursorToStringView);
+                info.availabilityZone = StringView(
```

Comment:
You have a ByteCursorToStringView function, shouldn't we use it here?

Reply:
We should, I forgot to make this change.